stustapay/tse/diebold_nixdorf_usb/handler.py

```diff
@@ -15,33 +16,58 @@
 class DieboldNixdorfUSBTSE(TSEHandler):
     def __init__(self, config: DieboldNixdorfUSBTSEConfig):
         self.websocket_url = config.diebold_nixdorf_usb_ws_url
+        self.websocket_timeout = config.ws_timeout
         self.background_task: typing.Optional[asyncio.Task] = None
         self.request_id = 0
         self.pending_requests: dict[int, asyncio.Future[dict]] = {}
+        self._started = asyncio.Event()
+        self._stop = False
+        self._ws = None
 
-    async def start(self):
-        self.background_task = asyncio.create_task(self.run())
+    async def start(self) -> bool:
+        self._stop = False
+        start_result: asyncio.Future[bool] = asyncio.Future()
+        self.background_task = asyncio.create_task(self.run(start_result))
+        return await start_result
 
     async def stop(self):
         # TODO cleanly cancel the background task
+        self._stop = True
         await self.background_task
+        self._started.clear()
 
-    async def run(self):
-        async with aiohttp.ClientSession() as session:
-            async with session.ws_connect(self.websocket_url) as ws:
-                receive_task = asyncio.create_task(self.recieve_loop(ws))
-                while True:
-                    await self.request("PingPong", ws=ws)
+    async def run(self, start_result: asyncio.Future[bool]):
+        async with contextlib.AsyncExitStack() as stack:
+            def set_ws(ws):
+                print(f'setting self._ws={ws}')
+                self._ws = ws
+            try:
+                session = await stack.enter_async_context(aiohttp.ClientSession(timeout=aiohttp.ClientTimeout(total=self.websocket_timeout, connect=self.websocket_timeout)))
+                LOGGER.info(f'connecting to DN TSE at {self.websocket_url!r}')
+                set_ws(await stack.enter_async_context(session.ws_connect(self.websocket_url)))
+            finally:
+                start_result.set_result(self._ws is not None)
+            stack.callback(set_ws, None)
 
-    async def request(self, command: str, *, ws, timeout: float = 5, **kwargs) -> dict:
+            receive_task = asyncio.create_task(self.recieve_loop())
+            await self.request("SetDefaultClientId", ClientID="DummyDefaultClientId")
+            while not self._stop:
+                print('sending pingpong')
+                result = await self.request("PingPong")

```

Comment:
this pingpong checks for service liveliness, right? what happens when it's no longer alive?

Reply:
repsonse timeout -> exception